packages/keystatic/src/component-blocks/cloud-image-preview.tsx

```diff
@@ -2,159 +2,410 @@
         typeof data.src === 'string'
       ) {
-        event.preventDefault();
-        props.onChange({
-          src: data.src,
-          alt: 'alt' in data && typeof data.alt === 'string' ? data.alt : '',
-          height: 'height' in data ? getDimension(data.height) : '',
-          width: 'width' in data ? getDimension(data.width) : '',
-        });
+        setState(cleanImageData(data));
         return;
       }
     } catch (err) {}
     const pattern = /^\s*!\[(.*)\]\(([a-z0-9_\-/:.]+)\)\s*$/;
     const match = text.match(pattern);
     if (match) {
-      event.preventDefault();
-      props.onChange({
-        src: match[2],
-        alt: match[1],
-        height: '',
-        width: '',
-      });
+      setState(
+        cleanImageData({
+          src: match[2],
+          alt: match[1],
+        })
+      );
       return;
     }
+    setState(cleanImageData({ src: text }));
   };
 
+  const src = state.src;
+
+  useEffect(() => {
+    if (!src) {
+      setStatus('');
+      return;
+    }
+    if (!isValidURL(src)) {
+      return;
+    }
+    setStatus('loading');
+    const img = new Image();
+    img.onload = () => {
+      const dimensions = {
+        width: img.width.toString(),
+        height: img.height.toString(),
+      };
+      setState(state => ({
+        ...state,
+        ...dimensions,
+      }));
+      setDimensions(dimensions);
+      setStatus('good');
+    };
+    img.onerror = () => {
+      setStatus('error');
+    };
+    img.src = src;
+    return () => {
+      img.onload = null;
+    };
+  }, [src]);
+
+  return (
+    <Dialog>
+      <Heading>
+        <Flex alignItems="center" gap="regular">
+          {image ? 'Edit' : 'Insert'} Cloud Image
+        </Flex>
+      </Heading>
+      <Header>
+        <Button
+          href={imageLibraryURL}
+          target="_blank"
+          rel="noreferrer"
+          prominence="low"
+          tone="accent"
+        >
+          <Text>Open Image Library</Text>
+          <Icon src={externalLinkIcon} />
+        </Button>
+      </Header>
+      <Content>
+        <Flex
+          elementType="form"
+          id="example-form"
+          onSubmit={e => {
+            e.preventDefault();
+            if (status !== 'good') return;
+            onChange(state);
+            onClose();
+          }}
+          direction="column"
+          gap="large"
+        >
+          <TextField
+            label="Image URL"
+            autoFocus
+            onPaste={onPaste}
+            onKeyDown={e => {
+              if (e.code === 'Backspace' || e.code === 'Delete') {
+                setState(cleanImageData());
+              }
+            }}
+            value={state.src}
+            description={
+              image
+                ? undefined
+                : 'Copy an Image URL from the Image Library and paste into this field to insert it.'
+            }
+            endElement={
+              status === 'loading' ? (
+                <div
```

Comment:
Any reason this doesn't use a layout primitive?

```js
<Flex height="regular" width="regular" alignItems="center" justifyContent="center" />
```
```

Reply:
I can't use the right tokens for `height` and `width` props that I need here, they're giving me TypeScript errors

Reply:
Aah, `element.regular`, got it :+1: